### PR TITLE
[5.1] Create Gate Access Admission, so that a Policy can deny access with a reason

### PR DESCRIPTION
**Pull request by JosephSilber**
This would allow the policy to return with a reason for denying access, which would then be used in the 403 `HttpException` in the `AuthorizesRequests` trait:

``` php
use App\User;
use App\Post;
use Illumiante\Auth\Access\CreatesGrants;

class PostPolicy
{
    use CreatesGrants;

    public function update(User $user, Post $post)
    {
        if ($user->id !== $post->user_id) {
            return $this->deny('You are not the owner of this post.');
        }

        if ($post->isClosed()) {
            return $this->deny('Post is closed and cannot be updated.');
        }

        return $this->allow();
    }
}
```

---

I'm not entirely set on the `Admission` nomenclature. I've played with `Grant`, `Pass`, `Card`, `Passage`, `Token` and more, but have found `Admission` to be the most semantic in this case. I'm open to suggestions.

### CODE REVIEW

**Comment by garygreen** [comment]
This is prefect, thanks a lot Joseph! :+1:

File: src/Illuminate/Auth/Access/Grant.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
+<?php
+
+namespace Illuminate\Auth\Access;
+
+class Grant
+{
+    /**
+     * Indicates is the request was granted.
```
Comment (garygreen) [inline src/Illuminate/Auth/Access/Grant.php]: if*

File: src/Illuminate/Auth/Access/Gate.php [hunk 2]
```diff
@@ -148,6 +150,38 @@ public function denies($ability, $arguments = [])
      */
     public function check($ability, $arguments = [])
```
Comment (garygreen) [inline src/Illuminate/Auth/Access/Gate.php]: Could this method be simplified as `return $this->getGrant($ability, $arguments)->allowed();` ?

Comment (JosephSilber) [inline src/Illuminate/Auth/Access/Gate.php]: @garygreen It could, but that would mean creating grants for no reason.

Comment (garygreen) [inline src/Illuminate/Auth/Access/Gate.php]: True

Comment (JosephSilber) [inline src/Illuminate/Auth/Access/Gate.php]: @garygreen the overhead of creating a simple grant object is negligible. I switched it out to what you proposed.

File: src/Illuminate/Auth/Access/Grant.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    protected $reason;
+
+    /**
+     * Create a new access grant instance.
+     *
+     * @return void
+     */
+    protected function __construct()
+    {
+        //
```
Comment (GrahamCampbell) [inline src/Illuminate/Auth/Access/Grant.php]: maybe include a comment here regarding the significance of making this method protected - can be useful for less experienced php developers reading laravel's source

Comment (JosephSilber) [inline src/Illuminate/Auth/Access/Grant.php]: I've never seen comments in Laravel's source targeted at "less experienced php developers".

I'll let @taylorotwell add it if he wants.

Comment (GrahamCampbell) [inline src/Illuminate/Auth/Access/Grant.php]: Hmm, let's say non-experts then. I recon many Laravel users will have never come across code like this before.

File: src/Illuminate/Foundation/Auth/Access/AuthorizesRequests.php [hunk 4]
```diff
@@ -63,14 +65,15 @@ protected function parseAbilityAndArguments($ability, $arguments)
...
-     * Throw an unauthorized exception based on gate results.
+     * Get an unauthorized exception based on gate results.
      *
      * @param  string  $ability
      * @param  array  $arguments
+     * @param  string  $reason
      * @return \Symfony\Component\HttpKernel\Exception\HttpException
      */
-    protected function createGateUnauthorizedException($ability, $arguments)
+    protected function createGateUnauthorizedException($ability, $arguments, $reason)
```
Comment (arrilot) [inline src/Illuminate/Foundation/Auth/Access/AuthorizesRequests.php]: This change breaks BC, doesn't it?

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Auth/Access/AuthorizesRequests.php]: technically, yes, but we'll probably get away with it

**Comment by patrickbrouwers** [comment]
Looks great! :)

File: src/Illuminate/Auth/Access/Admission.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    protected $reason;
+
+    /**
+     * Create a new admission instance.
+     *
+     * @return void
+     */
+    protected function __construct()
+    {
+        //
```
Comment (GrahamCampbell) [inline src/Illuminate/Auth/Access/Admission.php]: we need a comment here perhapse pointing out this method is protected to stop people sending PRs to remove a method they thought was useless

File: src/Illuminate/Auth/Access/Admission.php [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        return $admission;
+    }
+
+    /**
+     * Create a denied admission.
+     *
+     * @param  string  $reason
+     * @return static
+     */
+    public static function deny($reason = 'This action is unauthorized.')
```
Comment (GrahamCampbell) [inline src/Illuminate/Auth/Access/Admission.php]: why not use `null` as a default - php and symfony both do this as a convention for defaults, making it easy to skip args

**Comment by JosephSilber** [comment]
Now the other one is not just depressed, but lonely too 😛

**Pull request closed without merging**